src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,203 @@
+            IWaitIndicator waitService,
+            IVsOnlineServices vsOnlineServices,
+            IProjectThreadingService threadingService,
+            IVsUIService<IVsExtensibility, IVsExtensibility3> extensibility)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _projectVsServices = projectVsServices;
+            _workspace = workspace;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _roslynServices = roslynServices;
+            _waitService = waitService;
+            _vsOnlineServices = vsOnlineServices;
+            _threadingService = threadingService;
+            _extensibility = extensibility;
+        }
+
+        public void HandleRename(string oldFilePath, string newFilePath)
+        {
+            if (_oldFilePath == oldFilePath)
+            {
+                _isLanguageServiceDone = true;
+            }
+        }
+
+        public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
+        {
+            Assumes.Present(context);
+            Assumes.Present(node);
+            Assumes.Present(value);
+
+            _oldFilePath = node.FilePath;
+            string? newNameWithExtension = value;
+            // These variables are need to synchronize with Roslyn
+            _isLanguageServiceDone = false;
+
+            // Do not offer to rename the file in VS Online scenarios.
+            if (_vsOnlineServices.ConnectedToVSOnline)
+            {
+                return;
+            }
+
+            if (await IsAutomationFunction() || node.IsFolder)
+            {
+                // Do not display rename Prompt
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            // Get the list of possible actions to execute
+            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
+            CodeAnalysis.Project? project = GetCurrentProject();
+            if (project is null)
+            {
+                return;
+            }
+
+            CodeAnalysis.Document? oldDocument = GetDocument(project, _oldFilePath);
+            if (oldDocument is null)
+            {
+                return;
+            }
+
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, newNameWithExtension, oldDocument.Folders);
+
+            bool errorsDetected = false;
+            foreach (var action in documentRenameResult.ApplicableActions)
+            {
+                foreach (var error in action.GetErrors())
+                {
+                    errorsDetected = true;
+                }
+            }
+
+            // Check errors before applying changes
+            if (errorsDetected)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.RenameSymbolFailed, oldName);
```

Comment:
Its a bit odd that we give an error that rename failed before we ask them if they even want to rename.

Reply:
It should say "Cannot apply symbol rename"  ?